thalassa/api.py

```diff
@@ -65,24 +65,102 @@ def get_elevation_dmap(trimesh: gv.TriMesh, show_grid: bool = False) -> hv.Overl
         overlay = tiles * elevation
     return overlay
 
-def get_timeseries(fid,source=None):
-    fdata=fid._dataset
-    sxi,syi=fdata['SCHISM_hgrid_node_x'].data,fdata['SCHISM_hgrid_node_y'].data
-    def get_timeseries_xy(x,y):
-        dist=abs(sxi+1j*syi-x-1j*y); nid=np.nonzero(dist==dist.min())[0][0]
-        mti=fdata['time'].data; myi=fdata['elev'].data[:,nid].copy()
-        ha=hv.Curve((mti,myi),'time','elevation')
-        return ha
-
-    #fixed station
-    xys=((3.0,37),(30,33)) #todo: read fixed stations from inputs
-    hs=hv.HoloMap({'P_{}'.format(i+1):get_timeseries_xy(*xy) for i,xy in enumerate(xys)})
-
-    #dynamic
-    hd=hv.DynamicMap(get_timeseries_xy,streams=[PointerXY(x=0,y=0,source=source)])
-    hd=hd.opts(color='k',line_width=2,line_dash='dotted')
-
-    #points
-    hp=hv.HoloMap({'P_{}'.format(i+1):hv.Points(xy).opts(color='r') for i,xy in enumerate(xys)})
-
-    return hd*hs.overlay(),hp.overlay()
+#----------------------------------------------------------------------------------------
+#time series
+#----------------------------------------------------------------------------------------
+#define a class to store data related to time series points
+class timeseries_data:
+      def __init__(self):
+         self.init=False
+      def clear(self):
+          self.init=False
+
+#function for extracting time series@(x,y) from data
+def extract_timeseries(x,y,sx,sy,data):
+    dist=abs(sx+1j*sy-x-1j*y)
+    mdist=dist.min()
+    nid=np.nonzero(dist==mdist)[0][0]
+    mdata=data['elev'].data[:,nid].copy()
+    return mdist,mdata
+
+#function to dynamically add or remove pts by double clicking on the map
+def add_remove_pts(x,y,data,dataset,fmt):
+    if fmt=='add pts':
+       if len(data.xys)==0:
+          mdist,mdata=extract_timeseries(x,y,data.sx,data.sy,dataset)
+          hcurve=hv.Curve((data.time,mdata),'time','elevation')
+          if mdist<=data.mdist:
+             data.xys.append((x,y))
+             data.elev.append(mdata)
+             data.curve.append(hcurve)
+       else:
+          if data.xys[-1][0]!=x and data.xys[-1][1]!=y:
+             mdist,mdata=extract_timeseries(x,y,data.sx,data.sy,dataset)
+             hcurve=hv.Curve((data.time,mdata),'time','elevation')
+             if mdist<=data.mdist:
+                data.xys.append((x,y))
+                data.elev.append(mdata)
+                data.curve.append(hcurve)
+    elif fmt=='remove pts':
+       if len(data.xys)>0:
+          xys=np.array(data.xys)
+          dist=abs(xys[:,0]+1j*xys[:,1]-x-1j*y)
+          mdist=dist.min()
+          if mdist<=data.mdist:
+             nid=np.nonzero(dist==mdist)[0][0]
+             data.xys=[k for i,k in enumerate(data.xys) if i!=nid]
+             data.elev=[k for i,k in enumerate(data.elev) if i!=nid]
+             data.curve=[k for i,k in enumerate(data.curve) if i!=nid]
+    else:
+       pass
+
+#get time series plots
+def get_timeseries(source,data,dataset,ymin,ymax,fmt):
```

Comment:
The names of the arguments could probably be improved. E.g. 
- `data` is too generic. How about `ts_data` or something?
- `fmt` is probably some abbreviation. To me is it not apparent which one exactly. Generally speaking, it is usually a good idea not to make too many assumptions WRT who will read our code. Abbreviations, unless their meaning is 100% clear  from the context (e.g. physics formulas), should probably be avoided. Even then their meaning might not be clear to someone who is not a domain expert. To me, out of context, `fmt` can only mean `format`. In this case, if I understand what the code does, it holds the value of the `timeseries_pts` widget and indicates the "mode"  under which we are operating (i.e. `add_points`, `remove_points` or `clear`) which is not related to `format`. I would suggest finding something more specific.
- `source` is only used later on in the holoviews streams and it is a `gv.Trimesh` instance. I think that renaming that to `trimesh` would be both more consistent with the other functions in the `api.py` and would help the reader understand how he should use the `get_timeseries()` function.

Reply:
This is a generic function. Therefore, using generic name is more appropriate.  Later, data can be other instance. Also, source is not limited to Trimesh.